sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -158,25 +158,34 @@ case class AdaptiveSparkPlanExec(
       var result = createQueryStages(currentPhysicalPlan)
       val events = new LinkedBlockingQueue[StageMaterializationEvent]()
       val errors = new mutable.ArrayBuffer[SparkException]()
+      val runningStages = util.Collections.synchronizedSet(new util.HashSet[QueryStageExec]())
```

Comment:
one idea: if we only check "started stages", then this doesn't need to be thread-safe.

When we cancel stages, we only do it for  "started stages", and also exclude finished stages which `stage.resultOption.isDefined`.

Reply:
Ok, but we still need to exclude finished but failed stage(`Future.isFailure`) whose `resultOption` is empty in this way?

Maybe, we can assign `resultOption` to the exception for failed stages. Thus, we can also identify it as finished stage later.

Reply:
Can we update `QueryStageExec.cancel` to skip failed Future?

Reply:
I see. And check Future should also work for succeed stage? We can use the same way to skip all the finished stages(includes both succeed and failed).

Reply:
yup

Reply:
Wait, wait. After taking another look, I realized that `cancel` has already skipped finished stage. So, the problem, actually, only exists in stage which fails to initialize lazy val.